internal/services/sarif/sarif.go

```diff
@@ -0,0 +1,189 @@
+	}
+}
+
+func (s *Sarif) ConvertVulnerabilityToSarif() (report Report) {
+
+	// All runs in the report (each run is one tool's output)
+	report.Runs = []ReportRun{}
+
+	resultsByTool := make(map[string][]Result)
+
+	// Organize each run by its corresponding tool name
+	// Each run has only one tool providing values; since we run many tools, must have many runs
+	runsByTool := make(map[string]ReportRun)
+
+	// Organize rules by corresponding tool name, subkeying by RuleId
+	// Each tool has a unique list of rules with metadata
+	rulesByToolAndId := make(map[string]map[string]Rule)
+
+	// Organize artifacts for each run by corresponding tool name
+	// Each tool has a list of artifacts which are referred to by the results
+	artifactsByToolAndName := make(map[string]map[string]Artifact)
+
+	for index := range s.analysis.AnalysisVulnerabilities {
+		vTarget := &s.analysis.AnalysisVulnerabilities[index].Vulnerability
+		toolName := vTarget.SecurityTool.ToString()
+
+		// Test for first time seeing a given tool
+		if _, exists := runsByTool[toolName]; !exists {
+			rulesByToolAndId[toolName] = make(map[string]Rule)
+			artifactsByToolAndName[toolName] = make(map[string]Artifact)
+
+			reportRun := ReportRun{
+				Tool: s.newTool(vTarget),
+			}
+
+			// create the run and the tool
+			runsByTool[toolName] = reportRun
+
+			// add run to main (composite) report
+			report.Runs = append(report.Runs, reportRun)
+		}
+
+		// add result to run report
+		result := s.newResult(vTarget)
+		resultsByTool[toolName] = append(resultsByTool[toolName], result)
+
+		// add artifact to run report
+		artifact := s.newArtifact(vTarget)
+		artifactsByToolAndName[toolName][artifact.Location.Uri] = artifact
+
+		// add rule to tool
+		rule := s.newRule(vTarget)
+		rulesByToolAndId[toolName][rule.Id] = rule
+	}
+
+	for idx, runReport := range report.Runs {
+		toolName := runReport.Tool.Driver.Name
+		artifactMap := artifactsByToolAndName[toolName]
+		ruleMap := rulesByToolAndId[toolName]
+		resultMap := resultsByTool[toolName]
+
+		// Integrate artifacts and rules into the run map
+		// Using this intermediate map enforces uniqueness
+		for _, artifact := range artifactMap {
+			report.Runs[idx].Artifacts = append(report.Runs[idx].Artifacts, artifact)
+		}
+		for _, rule := range ruleMap {
+			report.Runs[idx].Tool.Driver.Rules = append(report.Runs[idx].Tool.Driver.Rules, rule)
+		}
+		for _, result := range resultMap {
+			report.Runs[idx].Results = append(report.Runs[idx].Results, result)
+		}
+	}
+
+	return report
+}
+
+func (s *Sarif) convertNonZeroIntStr(str string) int {
+	newInt, _ := strconv.Atoi(str)
+	if newInt > 0 {
+		return newInt
+	}
+	return 1
+}
+
+func (s *Sarif) newTool(vulnerability *vulnerability.Vulnerability) ScanTool {
+	return ScanTool{
+		Driver: ScanToolDriver{
+			Name:               vulnerability.SecurityTool.ToString(),
+			MoreInformationUri: "https://www.google.com", // TODO
+			Version:            "1.0.0",                  // TODO
```

Comment:
Since this information will not be filled I think that we should let empty.

Reply:
I agree with you in **concept** but the SARIF validator gives me complaints when I do this ( https://sarifweb.azurewebsites.net/Validation ):

`SARIF2005: runs[0].tool.driver: The tool 'Bandit' does not provide any of the version-related properties 'version', 'semanticVersion', 'dottedQuadFileVersion'. Providing version information enables the log file consumer to determine whether the file was produced by an up to date version, and to avoid accidentally comparing log files produced by different tool versions.`

`SARIF2005: runs[0].tool.driver: The tool 'Bandit' does not provide 'informationUri'. This property helps the developer responsible for addessing a result by providing a way to learn more about the tool.`

How do you want to handle this? Make the URL the Horusec doc page for the scanner? The Horusec root webpage?

Reply:
Hmmm, this is weird. What do you guys think? @nathanmartinszup @wiliansilvazup @iancardosozup

Reply:
One option could be to put the CLI version and link to [documentation](https://docs.horusec.io/docs/cli/analysis-tools/overview/) about the orchestrated tools, but I don't think it's ideal.

Reply:
we could add some fields to vulnerability object ( tool version and tool uri), but i think this should be made in another PR.  @nathanmartinszup idea looks good to me in a short term

Reply:
I can do this in short-term -- is there a place where I can pull version ID of the build to inject in here?

I'll also see if I can open a second PR (or third? where are we ...) around adding those fields.